### PR TITLE
Api V1: discuss proposal here (draft state)

### PR DESCRIPTION
**Pull request by rrnewton**
This is a WIP proposal.  Look at your own risk.  I'll let you know when it is in a "full draft" state.

[Rendered here](https://github.com/iu-parfunc/systrace/blob/apiv1/rfcs/shared_api_v1/src/main.rs).

### CODE REVIEW

File: rfcs/shared_api_v1/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+        match self {
+            IncrMsg(n) => n.serialize(serializer)
+        }
+    }        
+}
+
+impl SystraceTool for CounterTool {
+    type Glob = u64;    
+    type GlobMethodArgs = IncrMsg;
+    type GlobMethodResult = ();
```
Comment (ckoparkar) [inline rfcs/shared_api_v1/src/main.rs]: @rrnewton How can one inspect the global state 1) from a tool and 2) from the guest? If it's via `receive_rpc` to retrieve certain fields of a struct, or `u64` in this case, specifying `GlobMethodResult` might be tricky. Or can we have a `fetch_global_state () -> &Glob` ?

File: rfcs/shared_api_v1/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+    fn get_regs(&self) -> Regs;
+    fn get_static_config(&self) -> StaticConfig;
+    fn get_dynamic_config(&self) -> DynConfig;
+}
+
+// TODO: Full "Instrumentor" interface that allows sending global RPCs as well.
+pub trait Instrumentor : GuestAccess { 
+    /// Send an RPC message to wherever the global state is stored, 
+    /// synchronously block until a response is received.
+    fn send_rpc_sync(self : &mut Self, args : SerializedVal) -> SerializedVal;
```
[the remark this answers is not in the document]
Comment (rrnewton) [reply rfcs/shared_api_v1/src/main.rs]: Maybe?  It's only local intra-machine "RPC", so I was thinking that perhaps we do not need to expose a failure mode.  But maybe we do.

[the remark this answers is not in the document]
Comment (ckoparkar) [reply rfcs/shared_api_v1/src/main.rs]: Actually, a lot of things could be `Result`s. Minor point, and not important atm.

File: rfcs/shared_api_v1/src/main.rs (hunk 1 above)
Comment (rrnewton) [reply rfcs/shared_api_v1/src/main.rs]: Well, only the Glob RPC methods should be able to directly access the `Glob` type.  The `handle_event` callbacks should be forced to send RPCs, irrespective of whether they happen to execute in the guest address space or in the centralized tracer/daemon.

Comment (ckoparkar) [reply rfcs/shared_api_v1/src/main.rs]: Alright. If we decide to encode this as separate RPC's to retrieve specific fields of a struct, or `u64` in this case, specifying `GlobMethodResult` might be tricky? Or can we have a `fetch_Glob () -> &Glob`, which would be a RPC call, but would just return a read-only `Glob`.

**Comment by rrnewton** [comment]
I'd also like to get a review by @chamibuddhika at some point.  What's missing from this API proposal vis a vis Dyninst, etc?

**Comment by rrnewton** [comment]
Might as well store this in the `docs/` folder, which means going ahead and merging.  Lets add other docs there too.  Delete when it becomes irrelevant.

**Pull request merged**